pkg/relay/ethereum/tx.go

```diff
@@ -35,68 +36,19 @@ func (c *Chain) SendMsgs(msgs []sdk.Msg) ([]core.MsgID, error) {
 	}
 	logger := c.GetChainLogger()
 	var msgIDs []core.MsgID
-	for i, msg := range msgs {
-		logger := &log.RelayLogger{Logger: logger.With(
-			logAttrMsgIndex, i,
-			logAttrMsgType, fmt.Sprintf("%T", msg),
-		)}
 
-		opts, err := c.TxOpts(ctx, true)
-		if err != nil {
-			return nil, err
-		}
+	iter := NewCallIter(msgs, skipUpdateClientCommitment)
+	for !iter.End() {
+		from := iter.Cursor()
+		logger := &log.RelayLogger{Logger: logger.With(logAttrMsgIndexFrom, from)}
 
-		// gas estimation
-		{
-			logger := &log.RelayLogger{Logger: logger.Logger}
-
-			opts.GasLimit = math.MaxUint64
-			opts.NoSend = true
-			tx, err := c.SendTx(opts, msg, skipUpdateClientCommitment)
-			if err != nil {
-				logger.Error("failed to build tx for gas estimation", err)
-				return nil, err
-			}
-
-			if rawTxData, err := tx.MarshalBinary(); err != nil {
-				logger.Error("failed to encode tx", err)
-			} else {
-				logger.Logger = logger.With(logAttrRawTxData, hex.EncodeToString(rawTxData))
-			}
-
-			estimatedGas, err := c.client.EstimateGasFromTx(ctx, tx)
-			if err != nil {
-				if revertReason, rawErrorData, err := c.getRevertReasonFromEstimateGas(err); err != nil {
-					// Raw error data may be available even if revert reason isn't available.
-					logger.Logger = logger.With(logAttrRawErrorData, hex.EncodeToString(rawErrorData))
-					logger.Error("failed to get revert reason", err)
-				} else {
-					logger.Logger = logger.With(
-						logAttrRawErrorData, hex.EncodeToString(rawErrorData),
-						logAttrRevertReason, revertReason,
-					)
-				}
-
-				logger.Error("failed to estimate gas", err)
-				return nil, err
-			}
-
-			txGasLimit := estimatedGas * c.Config().GasEstimateRate.Numerator / c.Config().GasEstimateRate.Denominator
-			if txGasLimit > c.Config().MaxGasLimit {
-				logger.Warn("estimated gas exceeds max gas limit",
-					logAttrEstimatedGas, txGasLimit,
-					logAttrMaxGasLimit, c.Config().MaxGasLimit,
-				)
-				txGasLimit = c.Config().MaxGasLimit
-			}
-			opts.GasLimit = txGasLimit
-		}
-
-		opts.NoSend = false
-		tx, err := c.SendTx(opts, msg, skipUpdateClientCommitment)
+		tx, err := iter.SendTx(ctx, c)
+		logger.Logger = logger.With(logAttrMsgIndexTo, iter.Cursor())
 		if err != nil {
 			logger.Error("failed to send msg", err)
 			return nil, err
+		} else if tx == nil {
+			break
```

Comment:
It is not occured because when !iter.End() iter.SendTx() should returns err or non nil tx value.
But this spec is depends on implementation details.
It is kindfull to show explicitly that tx==nil does not break a program and you(code reader) need not consider tx==nil case.